# Topic: Assignment and Functions

## Python Concepts: 
* Variables
* Assignment Statements
* Defining new functions
* Returning value from function
* Parameters and arguments

### Keywords
* <code>def</code>
* <code>return</code>

## EXERCISES:    

### Variables and Assignment
&nbsp;&nbsp;&nbsp;&nbsp;Average Speed
<br/>
&nbsp;&nbsp;&nbsp;&nbsp;Earth vs Moon
<br/>
&nbsp;&nbsp;&nbsp;&nbsp;Microcentury

### Code Tracing
### User Defined Functions
&nbsp;&nbsp;&nbsp;&nbsp;Hot Air Balloon
<br/>
&nbsp;&nbsp;&nbsp;&nbsp;Traffic - Safety
<br/>
&nbsp;&nbsp;&nbsp;&nbsp;Distance Travelled
<br/>
&nbsp;&nbsp;&nbsp;&nbsp;Difference in Distance Travelled
<br/>
&nbsp;&nbsp;&nbsp;&nbsp;Road Curvature Design
### Function Call Control Flow

<a name="variables"></a>
<hr/>

# Variables and Assignment

<a name="average_speed">
<b>Average Speed</b></a>

In [2]:
# Author: Colin Fidge 2021
# Imagine that you get on your bicycle and travel from
# your home to QUT at 30 km/hr.  After a hard day's
# study you cycle home again more slowly at 20 km/hr.
#
# Quickly now, what is your average speed for the whole
# round trip?  Be careful - most people get this
# wrong!  (Thanks to Professor Julius Sumner Miller for
# this brainteaser.)
#
# To check the correct answer, complete the code below.
# We have chosen an arbitrary distance of 6 km between
# your house and the university but the result is the
# same regardless of the distance.

# Given values:
distance_from_home_to_uni = 6 # km
speed_from_home_to_uni = 30 # km/hr
speed_from_uni_to_home = 20 # km/hr

# Complete the following code by replacing the question marks:

time_to_get_to_uni = distance_from_home_to_uni / speed_from_home_to_uni # hours

time_to_get_home = distance_from_home_to_uni / speed_from_uni_to_home # hours

total_travelling_time = time_to_get_to_uni + time_to_get_home # hours

total_distance_travelled = distance_from_home_to_uni * 2 # km

speed_for_round_trip = total_distance_travelled / total_travelling_time # km/hr

speed_for_round_trip

24.0

<a name="earth_v_moon"></a>
<b>Earth vs Moon</b>

In [ ]:
# Which has more dry land, the earth or the moon?
# [Quiz question from Time magazine]
# Author: Colin Fidge (2021)

# Complete the following code by replacing the question marks:

import math

# First calculate the amount of dry land on the earth

# The surface area of the earth is 5.1 x 10^8 sq km
earths_surface_area = 5.1E8 # sq km

# 71% of the earth's surface is covered in water.
earths_water_area = earths_surface_area * 0.71

earths_dry_land = earths_surface_area - earths_water_area

# Now calculate the moon's surface area
# NB: There is no water on the moon!
# Remember that the surface area of a sphere is
# 4 times pi times r squared

# The diameter of the moon is 3,475 km
moons_radius = 3475 / 2 # km

# Change the following expression to calculate the moons dry land
# using the variable moons_radius
# Answer is: 37936694 

moons_dry_land = 4 * math.pi * moons_radius ** 2

# Which has more dry land, the earth or the moon?

print(moons_dry_land, earths_dry_land)

<a name="microcentury"></a>
<b>Microcentury</b>

In [ ]:
# When asked how long his lectures were, Professor
# Julius Sumner Miller usually answered "about a
# microcentury."  As an arithmetic exercise
# we can find out how many minutes in a microcentury
# by multiplying the minutes in an hour x hours in a day x
# days in a year x years in a century
# then dividing by 1 million ("micro" means one millionth).

# Some of the values are familiar to us without much thought:
minutes_per_hour = 60
hours_per_day = 24
days_per_year = 365 # ish
years_per_century = 100

# But others require a bit of calculation.
# For each of the remaining variables, change the question marks 
# to an appropriate expression that uses the variables
# above, and no numeric literals

minutes_per_day = minutes_per_hour * hours_per_day
minutes_per_century = minutes_per_day * days_per_year * years_per_century
microcentury = minutes_per_century / 1000000

microcentury

<a name="tracing"></a>
<hr/>

# Code Tracing

In [16]:
# Now trace the following code with a Trace Table.
# You will need a piece of paper with some columns for each variable.
# Check that the final value for each variable is the same as that
# produced by the Python code

# Scenario:
# Savings account attracts no fees for deposits or withdrawals.
# The Bank charges a fee of 1% per credit card transaction!
# Credit charge calculations are rounded to two decimal places

# opening balance of two bank accounts
savings_account = 1535.0 # credit
credit_account = 2997.56  # available credit 

# Some transactions on those accounts
# Deposit salary into savings acccount
salary = 2500
savings_account += salary # shortform of: savings_account = savings_account + salary

# Pay a bill using the credit account
rego_fees = 540.12
credit_charge = round(rego_fees * .01, 2)
credit_account = credit_account - (rego_fees + credit_charge)

# Transfer from savings to cover rego payment and charges
savings_account -= rego_fees + credit_charge # shortform of: savings_account = savings_account - (rego_fees + credit_charge)

print(savings_account, credit_account)

3489.48 2452.04


<table>
<tr><td><td></td></td><td><td>savings_account</td></td><td><td>credit_account</td></td><td><td>salary</td></td><td><td>rego_fees</td></td><td><td>credit_charge</td></td></tr>
<tr><td><td>savings_account = 1535.0</td></td><td><td>1535</td></td><td><td></td></td><td><td></td></td><td><td></td></td><td><td></td></td></tr>
<tr><td><td>credit_account = 2997.56</td></td><td><td>1535</td></td><td><td>2997.56</td></td><td><td></td></td><td><td></td></td><td><td></td></td></tr>
<tr><td><td>salary = 2500</td></td><td><td>1535</td></td><td><td>2997.56</td></td><td><td>2500</td></td><td><td></td></td><td><td></td></td></tr>
<tr><td><td>savings_account += salary</td></td><td><td>4035</td></td><td><td>2997.56</td></td><td><td>2500</td></td><td><td></td></td><td><td></td></td></tr>
<tr><td><td>rego_fees = 540.12</td></td><td><td>4035</td></td><td><td>2997.56</td></td><td><td>2500</td></td><td><td>540.12</td></td><td><td></td></td></tr>
<tr><td><td>credit_charge = round(rego_fees * .01, 2)</td></td><td><td>4035</td></td><td><td>2997.56</td></td><td><td>2500</td></td><td><td>540.12</td></td><td><td>5.40</td></td></tr>
<tr><td><td>credit_account = credit_account - (rego_fees + credit_charge)</td></td><td><td>4035</td></td><td><td>2452.04</td></td><td><td>2500</td></td><td><td>540.12</td></td><td><td>5.40</td></td></tr>
<tr><td><td>savings_account -= rego_fees + credit_charge</td></td><td><td>3489.48</td></td><td><td>2452.04</td></td><td><td>2500</td></td><td><td>540.12</td></td><td><td>5.40</td></td></tr>
</table>

<div style="background-color:pink">
Now trace the above code using the Jupyter Debugger.
</div>

# User Defined Functions


<a name="hot_air_balloon"><b>Hot Air Balloon
    </b></a>

In [ ]:
# Write a function to calculate the maximum weight a hot air balloon of
# negligible mass can carry given:
# a radius (in metres) and the outside and inside air densities

# Use the Buoyancy force equation:
#     F = ρV g
# where:
#    F is the Buoyancy force
#    ρ is the density difference
#    V is the displaced volume (4/3 πr^3) and 
#    g is the local gravitational constant (normally assumed to be equal to 9.81).

# Using Newton’s third law; we find that the maximum weight it could carry
# equals the lifting force of the balloon. Therefore: F_max = ρV g

def hot_air_balloon_capacity(radius, outside_density, inside_density) :
    V = 4/3 * math.pi * radius ** 3
    g = 9.81
    F = (outside_density - inside_density) * V * g
    return F

# Test 1 - outside air density is 1.2 kg.m^-3; inside helium is 0.2 kg.m^-3
hot_air_balloon_capacity(10, 1.2, 0.2) 
# should return 41092.03190895449

In [ ]:
# Test 2 - outside air density is 1.2 kg.m^-3; inside hydrogen is 0.0899 kg.m^-3
hot_air_balloon_capacity(10, 1.2, 0.0899)
# should return 45616.264622130366

<a name="transport_safety"><a>
## Transport - Safety


Formulas:
    <ul><li>distance travelled = $\frac{5}{18} sr$
    </li>
    </ul>
Where:
* s = speed
* r = reaction time

<hr>


<a name="distance_travelled"><b>Distance Travelled</b></a>

In [ ]:
# Time taken by a driver of a vehicle to react to a situation is termed as 
# drivers’ reaction time. 

# Define a function to estimate the distance (in metres)
# travelled by the vehicle before driver initiates 
# a physical response to a situation (such as commence breaking).
def distance_travelled(speed, reaction_time):
    return  (5/18) * speed * reaction_time
    
distance_travelled(100, 1.2)
# should return 33.333333333333336


<a name="distance_travelled_difference"><b>Difference in Distance Travelled
    </b></a>

In [ ]:
# Driving under the influence of alcohol has a significant impact on drivers’ driving ability.
# Say a driver has a normal reaction time of 1.2 seconds and his/her reaction time
# increases to 5 seconds for driving under the influence of alcohol. 
# A driver is diving at a speed of 100 km/hr and is unfortunately under the influence 
# of alcohol. Compared to a normal situation, how much extra the vehicle will move 
# before driver initiates a physical response to the situation?

# Use the distance_travelled function again to find the difference.
distance_travelled(100, 5) - distance_travelled(100, 1.2)

<hr>
<a name="road_curvature"><b>Road Curvature Design
    </b></a>

minimum radius of curvature: $R = \frac{S^2}{127 . (f + e)}$
where:
* R = minimum radiusc of curvature in meters
* S = speed in km/hr
* f = coefficient of side friction on the road 
* e = superelevation expressed as a decimal


In [ ]:
# Roads are not perfectly straight and do have curves in the horizontal plane. 
# For safe driving, the horizontal curvature of the road should have minimum radius 
# (termed as minimum radius of curvature) and it is expressed as the equation shown above.
            
# Write a Python function to estimate R, given S, f and e.

def radius_of_curvature(speed, friction, superelevation):
    return (speed ** 2) / (127 * (friction + superelevation)) 

radius_of_curvature(80, 0.14, 0.08) 
# should return 229.06227630637076

<a name="trace_back"></a>
<hr/>

# Function Call Control Flow


In [ ]:
# This exercise is designed to illustrate the order and flow of execution
# of Python code that involves function calls.  The code below
# calculates the number of carpet tiles required to completely
# cover a floor in a building. However, it contains a logical error... 

import math

def get_tiles_in_one_direction(overall_length, tile_length):
    return math.ceil(overall_length / tile_length)

# Given the dimensions of a room and the size of a single carpet tile,
# calculate the number of tiles required to cover the entire room (assuming
# no space is required between the tiles).
def get_tiles_required(room_width, room_length, tile_size):
    return get_tiles_in_one_direction(room_width, tile_size) * get_tiles_in_one_direction(room_length, tile_size)

# Here are the dimensions of the room and tiles
room_width =  5.1
room_length = 3.3
tile_size = 0.9

# Calculate and display the number of tiles required to completely
# cover the floor
get_tiles_required(room_width, room_length, tile_size)

# As you will see when you execute this code, it generates a runtime error.  

# Looking at the code, list the function calls in the order they would be made by Python.

# Now look at the traceback information that is displayed below after you run the code.
# It is obvious that an error has occurred because the output is highlighted in a striking colour!
# The traceback indicates the flow of control from top to bottom - that is, the last --->
# that you see in the traceback indicates the line of code where the exception (error) is generated,
# and the previous ---> in the traceback is the call that is made prior to that.

# You may also use the debugger and/or the post morteim debugger to help diagnose the problem if necessary.

# Now rectify the code where the error originated by removing an unnecessary and illogical, but aptly named function call.